### PR TITLE
Add Binary Cross Entropy loss

### PR DESCRIPTION
**Pull request by m0saan**
## Summary
This pull request introduces the implementation of Binary Cross Entropy (BCE) Loss as both a standalone function and a class.

## Changes Made
1. **Added BCE Loss Function:**
   - Implemented the `binary_cross_entropy` function, which calculates the binary cross entropy loss given predicted probabilities and target labels.

2. **Added BCE Loss Class:**
   - Introduced the `BCELoss` class, which encapsulates the BCE loss functionality.

3. **Documentation:**
   - Provided comprehensive docstrings for the new function and class.
`   - Included usage examples to guide users on how to apply the BCE loss in their projects.

## Related Issues
Closes [#76] 

## Checklist
- [x] Code follows the project's style guidelines.
- [ ] Added appropriate unit tests.
- [x] Documentation has been updated to reflect the changes.
- [x] Before submitting this PR, check the [contribution guidelines](CONTRIBUTING.md).

- [x] Make sure your code is formatted: `pre-commit run --all-files`.

### CODE REVIEW

File: python/mlx/nn/losses.py [hunk 1]
```diff
@@ -24,7 +35,42 @@ def cross_entropy(
...
+    Returns:
+        mx.array: The computed binary cross entropy loss.
+    Examples:
+        >>> import mlx.core as mx
+        >>> import mlx.nn as nn
+        >>> inputs = mx.array([0.1, 0.2, 0.3, 0.4])
+        >>> targets = mx.array([0, 0, 1, 1])
+        >>> loss = nn.losses.binary_cross_entropy(inputs, targets)
+        >>> loss
+        array([0.612192])
```
Comment (awni) [inline python/mlx/nn/losses.py:59]: ❤️

File: python/mlx/nn/losses.py [hunk 2]
```diff
@@ -24,7 +35,42 @@ def cross_entropy(
...
+    """
+    loss = -targets * mx.log(inputs) - (1 - targets) * mx.log(1 - inputs)
+
+    return _reduce(loss, reduction)
+
+
+@_make_loss_module(binary_cross_entropy)
+class BCELoss(Module):
+    pass
+
```
Comment (awni) [inline python/mlx/nn/losses.py:69]: Cool! Could you add a test using the module as well, just to make sure the "module making" code works?

File: python/mlx/nn/losses.py [hunk 3]
```diff
@@ -24,7 +35,42 @@ def cross_entropy(
...
 
-def l1_loss(predictions: mx.array, targets: mx.array, reduction: str = "none") -> mx.array:
+def binary_cross_entropy(
+    inputs: mx.array, targets: mx.array, reduction: str = "none"
+) -> mx.array:
+    """
+    Computes the binary cross entropy loss between inputs and targets.
+
+    Args:
+        inputs (mx.array): The predicted inputs.
```
Comment (awni) [inline python/mlx/nn/losses.py:45]: Would be good to clarify this loss expects the post sigmoid inputs / inputs should be probabilities.

**Review by awni — CHANGES_REQUESTED** [review]
This is awesome, thanks! Just a couple minor comments, then we can merge!

Also could you add the new loss to the docs? https://github.com/ml-explore/mlx/blob/main/docs/src/python/nn.rst#loss-functions

File: python/mlx/nn/losses.py (hunk 1 above)
Comment (m0saan) [reply python/mlx/nn/losses.py:59]: Thank you so very much for your excellent input!

File: python/mlx/nn/losses.py (hunk 2 above)
Comment (m0saan) [reply python/mlx/nn/losses.py:69]: ofc, on it!

**Comment by m0saan** [comment]
> This is awesome, thanks! Just a couple minor comments, then we can merge!
> 
> Also could you add the new loss to the docs? https://github.com/ml-explore/mlx/blob/main/docs/src/python/nn.rst#loss-functions

done! :p

**Review by awni — APPROVED** [review]
Awesome, thanks!!

**Comment by awni** [comment]
FYI looks like you need to rebase before I can merge.

**Comment by awni** [comment]
@m0saan I still see a conflict with `python/mlx/nn/losses.py`. If you pull and rebase on main, resolve any conflicts, then force push it should be ok to merge.

**Comment by m0saan** [comment]
> @m0saan I still see a conflict with `python/mlx/nn/losses.py`. If you pull and rebase on main, resolve any conflicts, then force push it should be ok to merge.

now looks good to me, thanks!

**Review by awni — APPROVED** [review]
Perfect, thanks!
